install-deps.sh

```diff
@@ -469,44 +470,51 @@ function activate_virtualenv() {
     . $env_dir/bin/activate
 }
 
+function preload_wheels_for_tox() {
+    local ini=$1
+    shift
+    pushd .
+    cd $(dirname $ini)
+    local require_files=$(ls *requirements*.txt 2>/dev/null) || true
+    local constraint_files=$(ls *constraints*.txt 2>/dev/null) || true
+    local require=$(echo -n "$require_files" | sed -e 's/^/-r /')
+    local constraint=$(echo -n "$constraint_files" | sed -e 's/^/-c /')
+    local md5=wheelhouse/md5
+    if test "$require"; then
+        if ! test -f $md5 || ! md5sum -c $md5 > /dev/null; then
+            rm -rf wheelhouse
+        fi
+    fi
+    if test "$require" && ! test -d wheelhouse ; then
+        for interpreter in python2.7 python3 ; do
+            type $interpreter > /dev/null 2>&1 || continue
+            activate_virtualenv $top_srcdir $interpreter || exit 1
+            populate_wheelhouse "wheel -w $wip_wheelhouse" $require $constraint || exit 1
+        done
+        mv $wip_wheelhouse wheelhouse
+        md5sum $require_files $constraint_files > $md5
+    fi
+    popd
+}
+
 # use pip cache if possible but do not store it outside of the source
 # tree
 # see https://pip.pypa.io/en/stable/reference/pip_install.html#caching
-mkdir -p install-deps-cache
-top_srcdir=$(pwd)
-export XDG_CACHE_HOME=$top_srcdir/install-deps-cache
-wip_wheelhouse=wheelhouse-wip
+if $for_make_check; then
```

Comment:
As I cannot exactly tell what `for_make_check` is for, and how relates to pre-populating cache, would it make sense to have an intermediate mapping:
```sh
if $whatever_condition; then
  for_make_check=1
fi

#.. 

if $for_make_check; then
  prepopulate_python=1
  # or even more clear?
  offline_build=1
fi

if $prepopulate_python / $offline_build; then
...
fi
```

Reply:
@epuertat i think collecting the wheels beforehand also helps to reduce the traffic and shorten the time for running test.

Reply:
will add a comment to explain how/why "make check" is connected to python wheels.

Reply:
> @epuertat i think collecting the wheels beforehand also helps to reduce the traffic and shorten the time for running test.

IMHO if that were the only reason (and not doing offline builds), pip caching should work ok. Pre-fetching here wouldn't work that ok if you just wanted to build a single Ceph component (e.g: Dashboard) with Python3 only.

Reply:
pip caching cannot cache the pre-built binary wheels, i think. what it caches is the pre-cooked binary / source packages.